platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractMinimalProvider.java

```diff
@@ -169,12 +171,15 @@ public ProviderInternal<T> asSupplier(DisplayName owner, Class<? super T> target
     private static String formatInvalidTypeException(String owner, Class<?> targetType, Class<?> selfType) {
```

Comment:
:x: This check isn't only applied to shared build services, but to any property/provider. Having the `@ServiceReference` suggestion is going to confuse people. I think we can push the suggestion logic down into `BuildServiceProvider` instead.